src/client/components/UserPage/Drawer/ControlPanel/widgets/DownloadButton.tsx

```diff
@@ -6,12 +6,49 @@ import {
   MenuItem,
   Typography,
 } from '@material-ui/core'
+import FileSaver from 'file-saver'
 
 import TrailingButton from './TrailingButton'
 import downloadIcon from '../assets/download-icon.svg'
-import { useDrawerState, useDrawerDispatch } from '../..'
-import QRCodeModal from '../QRCodeModal'
-import DrawerActions from '../helpers/reducers'
+import { useDrawerState } from '../..'
+import ImageFormat from '../../../../../../shared/util/imageFormat'
+import { postJson } from '../../../../../util/requests'
+
+// Gets file extension from content-type.
+function getFileExtension(format: ImageFormat) {
+  switch (format) {
+    case ImageFormat.SVG:
+      return 'svg'
+    case ImageFormat.PNG:
+      return 'png'
+    case ImageFormat.JPEG:
+      return 'jpeg'
+    default:
+      throw Error('Invalid format passed to getFileExtension')
+  }
+}
+
+// Downloads QR code from server.
+async function downloadServerQrCode(
+  shortLink: string,
+  format: ImageFormat,
+): Promise<void> {
+  if (typeof window === 'undefined') {
+    throw Error('This code should only run in a browser environment')
+  }
+  const url = `https://${document.location.host}/${shortLink}`
```

Comment:
Curious to know your reasoning behind letting the client decide the whole url rather than just specifying shortlink and having the server generate the rest

Reply:
The server now returns a proposed filename in the headers (since body and content-type will be reserved for the image file), which the client will use when saving the image. If the filename header does not exist, the client fallbacks to the short url generated.